src/canvas/canvas_events.ts

```diff
@@ -0,0 +1,1725 @@
+    functor(canvasElement, `${eventTypePrefix}enter`, this._onMouseEnter);
+    functor(canvasElement, 'wheel', this._onMouseWheel);
+    functor(canvasElement, 'contextmenu', this._onContextMenu);
+    functor(canvasElement, 'dblclick', this._onDoubleClick);
+    functor(canvasElement, 'dragstart', this._onDragStart);
+    functor(canvasElement, 'dragend', this._onDragEnd);
+    functor(canvasElement, 'dragover', this._onDragOver);
+    functor(canvasElement, 'dragenter', this._onDragEnter);
+    functor(canvasElement, 'dragleave', this._onDragLeave);
+    functor(canvasElement, 'drop', this._onDrop);
+    if (!this.enablePointerEvents) {
+      functor(canvasElement, 'touchstart', this._onTouchStart, addEventOptions);
+    }
+    // if (typeof eventjs !== 'undefined' && eventjsFunctor in eventjs) {
+    //   eventjs[eventjsFunctor](canvasElement, 'gesture', this._onGesture);
+    //   eventjs[eventjsFunctor](canvasElement, 'drag', this._onDrag);
+    //   eventjs[eventjsFunctor](
+    //     canvasElement,
+    //     'orientation',
+    //     this._onOrientationChange
+    //   );
+    //   eventjs[eventjsFunctor](canvasElement, 'shake', this._onShake);
+    //   eventjs[eventjsFunctor](canvasElement, 'longpress', this._onLongPress);
+    // }
+  }
+
+  /**
+   * Removes all event listeners
+   */
+  removeListeners() {
+    this.addOrRemove(removeListener, 'remove');
+    // if you dispose on a mouseDown, before mouse up, you need to clean document to...
+    const eventTypePrefix = this._getEventPrefix();
+    removeListener(
+      fabric.document,
+      `${eventTypePrefix}up`,
+      this._onMouseUp as EventListener
+    );
+    removeListener(
+      fabric.document,
+      'touchend',
+      this._onTouchEnd as EventListener,
+      addEventOptions
+    );
+    removeListener(
+      fabric.document,
+      `${eventTypePrefix}move`,
+      this._onMouseMove as EventListener,
+      addEventOptions
+    );
+    removeListener(
+      fabric.document,
+      'touchmove',
+      this._onMouseMove as EventListener,
+      addEventOptions
+    );
+  }
+
+  /**
+   * @private
+   */
+  private _bindEvents() {
+    if (this.eventsBound) {
+      // for any reason we pass here twice we do not want to bind events twice.
+      return;
+    }
+    (
+      [
+        '_onMouseDown',
+        '_onTouchStart',
+        '_onMouseMove',
+        '_onMouseUp',
+        '_onTouchEnd',
+        '_onResize',
+        // '_onGesture',
+        // '_onDrag',
+        // '_onShake',
+        // '_onLongPress',
+        // '_onOrientationChange',
+        '_onMouseWheel',
+        '_onMouseOut',
+        '_onMouseEnter',
+        '_onContextMenu',
+        '_onDoubleClick',
+        '_onDragStart',
+        '_onDragEnd',
+        '_onDragProgress',
+        '_onDragOver',
+        '_onDragEnter',
+        '_onDragLeave',
+        '_onDrop',
+      ] as const
+    ).forEach((eventHandler) => {
```

Comment:
I think this an inversion of cause and effect and i don't want to do it.
Fat arrow do not have a THIS and so in order to be used in a class they get bound at constructor time, otherwise when you call class.method() you have no idea which the THIS is if you don't bind it.

So what happens is that you need to bind them to even work in a class, and so as a side effect, being bound, they work as event handlers.

Once those are written down as fat arrow, in 3 months someone wants to convert them to function, there is no written anywhere that those should be bound again.

In our case those methods are supposed to be event handlers, so they need to be bound to the class because they will be executed out of context, and so we bind them explicitly.

MDN is clear on the fact that arrow functions are not good for classes: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions and the short hand syntax indeed make them not even more concise.

Those are useful exactly when you need an outside this for your expression ( forEach loops, map loops and so on ).

Reply:
It says exactly what I wrote. It creates auto bound methods

![image](https://user-images.githubusercontent.com/34343793/209908998-79fe619d-742c-4ec4-ad50-3c9d50226abe.png)
![image](https://user-images.githubusercontent.com/34343793/209909039-c4591c5f-5e35-4c77-909f-48e432d9043e.png)